spec.bs

```diff
@@ -247,35 +260,55 @@ A user agent holds an <dfn>aggregatable report cache</dfn>, which is a [=list=]
+1. Let |reportingOrigin| be |context|'s [=environment settings object/origin=].
 1. [=Assert=]: |reportingOrigin| is a [=potentially trustworthy origin=].
-1. Let |path| be the [=string/concatenation=] of «"<code>.well-known/private-aggregation/report-</code>", |api|».
-1. Let |base| be the result on running the [=URL parser=] on the [=serialization of an origin|serialization=] of |reportingOrigin|.
-1. [=Assert=]: |base| is not failure.
-1. Let |result| be the result of running the [=URL parser=] on |path| with |base|.
-1. [=Assert=]: |result| is not failure.
-1. Return |result|.
+1. If the user agent's [=aggregatable report cache=] [=set/size=] is larger than [=maximum stored reports=], return.
+1. Let |truncatedContributions| be a new [=list/is empty|empty=] [=list=].
+1. If |contributions| has a [=list/size=] greater than [=maximum report contributions=]:
+    1. [=set/For each=] |n| of [=the exclusive range|the range=] 0 to [=maximum report contributions=], exclusive:
+        1. [=set/Append=] |contributions|[|n|] to |truncatedContributions|.
+1. Otherwise, set |truncatedContributions| to |contributions|.
+1. Let |contributionSum| be 0.
+1. [=set/iterate|For each=] |contribution| of |truncatedContributions|:
+    1. [=Assert=]: |contribution|["|value|"] is non-negative.
+    1. Add |value| to |contributionSum|.
+1. Let |sufficientBudget| be the result of [=PrivateAggregation/consuming budget if permitted=] given |contributionSum|, |reportingOrigin|, |api| and |context|'s [=current wall time=].
+1. If |sufficientBudget| is false, return.
+1. Let |report| be the result of [=PrivateAggregation/obtaining an aggregatable report=] given |reportingOrigin|, |api|, |truncatedContributions| and |context|'s [=current wall time=].
+1. [=set/Append=] |report| to the user agent's [=aggregatable report cache=].
+
+Issue: Do we need to ensure the reports aren't queued after being sent?
 
-Issue: Remove export when this definition is used.
+Issue: Do we need to address user settings here at all?
 
-To <dfn for="PrivateAggregation">obtain an aggregatable report</dfn> given an [=origin=] |reportingOrigin|, a [=context type=] |api| and a [=list=] of {{PAHistogramContribution}} |contributions|, perform the following steps. They return an [=aggregatable report=].
+Issue: Does the [=maximum stored reports=] part match the implementation?
+
+To <dfn algorithm for="PrivateAggregation">consume budget if permitted</dfn> given a {{long}} |value|, an [=origin=] <var ignore=''>origin</var>, a [=context type=] |api| and a [=moment=] |currentTime|, perform [=implementation-defined=] steps. They return a [=boolean=], which indicates whether there is sufficient 'contribution budget' left to send the requested contribution |value|. This budget should be bound to usage over time, e.g. the contribution sum over the last 24 hours. The algorithm should assume that the contribution will be sent if and only if true is returned.
```

Comment:
This algorithm doesn't really consume the budget. Maybe call it "check if there is sufficient budget"?

Reply:
The idea is that this algorithm does actually consume the budget, but leaves what that means as implementation-defined. Tried to reword slightly to clarify, but lmk what you think

Reply:
I see. Yeah it's clearer with the rewording, thanks.